src/common/notifications.cpp

```diff
@@ -315,3 +319,99 @@ void notifications::show_toast_with_activations(std::wstring message, std::wstri
+{
+    // DO NOT LOCALIZE any string in this function, because they're XML tags and a subject to
+    // https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/toast-xml-schema
+
+    if (!params.progress.has_value())
+    {
+        return;
+    }
+
+    std::wstring toast_xml;
+    toast_xml.reserve(2048);
+    std::wstring title{ L"PowerToys" };
+    if (winstore::running_as_packaged())
+    {
+        title += L" (Experimental)";
+    }
+
+    toast_xml += LR"(<?xml version="1.0"?><toast><visual><binding template="ToastGeneric"><text>)";
+    toast_xml += title;
+    toast_xml += L"</text><text>";
+    toast_xml += message;
+    toast_xml += L"</text>";
+    toast_xml += LR"(<progress title=")";
+    toast_xml += localized_strings::PT_UPDATE;
+    toast_xml += L" ";
+    toast_xml += version;
+    toast_xml += LR"(" value="{progressValue}" valueStringOverride="{progressValueString}" status="{progressStatus}"/>)";
+    toast_xml += L"</binding></visual></toast>";
+
+    XmlDocument toast_xml_doc;
+    xml_escape(toast_xml);
+    toast_xml_doc.LoadXml(toast_xml);
+    ToastNotification notification{ toast_xml_doc };
+
+    float progress = params.progress.value();
+    winrt::Windows::Foundation::Collections::StringMap map;
+    map.Insert(L"progressValue", std::to_wstring(progress));
+    map.Insert(L"progressValueString", std::to_wstring(static_cast<int>(progress * 100)) + std::wstring(L"%"));
+    map.Insert(L"progressStatus", localized_strings::DOWNLOAD_IN_PROGRESS);
+    winrt::Windows::UI::Notifications::NotificationData data(map);
+    notification.Data(data);
+    
+    const auto notifier = winstore::running_as_packaged() ? ToastNotificationManager::ToastNotificationManager::CreateToastNotifier() :
+                                                            ToastNotificationManager::ToastNotificationManager::CreateToastNotifier(WIN32_AUMID);
+
+    // Set a tag-related params if it has a valid length
+    if (params.tag.has_value() && params.tag->length() < 64)
+    {
+        notification.Tag(*params.tag);
+        if (!params.resend_if_scheduled)
+        {
+            for (const auto& scheduled_toast : notifier.GetScheduledToastNotifications())
+            {
+                if (scheduled_toast.Tag() == *params.tag)
+                {
+                    return;
+                }
+            }
+        }
+    }
+
+    notifier.Show(notification);
+}
+
+void notifications::update_progress_bar_toast(std::wstring plaintext_message, toast_params params)
+{
+    if (!params.progress.has_value())
+    {
+        return;
+    }
+
+    const auto notifier = winstore::running_as_packaged() ? ToastNotificationManager::ToastNotificationManager::CreateToastNotifier() :
+                                                            ToastNotificationManager::ToastNotificationManager::CreateToastNotifier(WIN32_AUMID);
+    float progress = params.progress.value();
+    if (progress > 1)
```

Comment:
Let's [clamp](https://en.cppreference.com/w/cpp/algorithm/clamp) the value in [0,1].